neutonm/might-and-magic-amber-island-mod
# Refactor: Mercenary Mechanics (+loads merc data from data/tables)

### Description:

_The old system was a bit sloppy and required general refactoring. Plus it wasn't friendly from architectural perspective.
The new version is more abstract/template-based + loads data from data tables._

### Issue:

https://github.com/neutonm/might-and-magic-amber-island-mod/issues/11

### Changes:

Note:
- _Last commit should be "append" (to last commit) but it got fucked up._

Bug fixes:
- Fixed "Game crashes at loading New Game's Autosave" bug (thanks to grayface's notes) (related to merc-api)
- Issue with not-reset summon charges is fixed (it could happen during travels to arena or changing dungeons)

API:
- Filename changes + removed Warder/Mercenary mercenary templates
- Cleaned/Refactored Merc "API" and conformed it to a functional programming style (<CLASS>_<FUNC>, e.g: Merc_Hire())
- Mercs provide more detailed information about their class/stats via "Information" dialogue option
- Abstracted Guildmaster:
- - Rehire feature: For a separate, reduced price
- - Browse Mercenaries feature: Guildmaster offers a list of discovered mercenaries 
- Data Tables utilization: 
- - Mercenary core data is loaded from Data/Tables/Mercs.txt
- - Mercenary credentials data is loaded from Data/Tables/MercsCredentials.txt)
- Couple of QoL tweaks such as no greeting message if merc is dead, etc.

Misc:
- New debug tweaks: Party starts with god() and 999999 gold when Game.Debug == true
- - additionally unlocks all mercenaries

### QA:

Default:
- Start with "Game.Debug = true"
- Open up console (F6) and type "evt.SpeakNPC(505)" (ctrl+enter)

Checks:
- Hire 
- Fight (Summon Merc) / Dismiss
- Fire then ReHire
- Make some Upgrades for each merc
- Merc dies from monsters
- Merc dies from party
- Loose mercs in dungeon and Recall them via GUildmaster in other dungeon/map
- Timers: Deplete summoning charges (note: Mercs summon charges are reset each day at 1 AM)
- - rest for 24h, confirm charge restoration
- - deplete charges in X dungeon, rest in Y dungeon/map
- - deplete charges in Amber Island + goto Arena and return immediately back
- Repeat some basic steps with save/loading
